src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleasePermissionService.cs

```diff
@@ -0,0 +1,90 @@
+using GovUk.Education.ExploreEducationStatistics.Admin.Services.Interfaces;
+using GovUk.Education.ExploreEducationStatistics.Admin.Services.Interfaces.Security;
+using GovUk.Education.ExploreEducationStatistics.Admin.ViewModels;
+using GovUk.Education.ExploreEducationStatistics.Common.Model;
+using GovUk.Education.ExploreEducationStatistics.Common.Services.Interfaces.Security;
+using GovUk.Education.ExploreEducationStatistics.Common.Utils;
+using GovUk.Education.ExploreEducationStatistics.Content.Model;
+using GovUk.Education.ExploreEducationStatistics.Content.Model.Database;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+
+namespace GovUk.Education.ExploreEducationStatistics.Admin.Services
+{
+    public class ReleasePermissionService : IReleasePermissionService
+    {
+        private readonly IPersistenceHelper<ContentDbContext> _persistenceHelper;
+        private readonly ContentDbContext _contentDbContext;
+        private readonly IUserService _userService;
+
+        public ReleasePermissionService(
+            IPersistenceHelper<ContentDbContext> persistenceHelper,
+            ContentDbContext contentDbContext,
+            IUserService userService)
+        {
+            _persistenceHelper = persistenceHelper;
+            _contentDbContext = contentDbContext;
+            _userService = userService;
+        }
+
+        public async Task<Either<ActionResult, List<ReleaseContributorViewModel>>> GetReleaseContributors(Guid releaseId)
+        {
+            return await _persistenceHelper
+                .CheckEntityExists<Release>(releaseId,
+                    query =>
+                        query.Include(r => r.Publication))
+                .OnSuccessDo(release => _userService
+                    .CheckCanUpdatePublicationReleaseRole(
+                        new Tuple<Publication, ReleaseRole>(release.Publication, ReleaseRole.Contributor)))
+                .OnSuccess(async release =>
+                {
+                    var allReleases = await _contentDbContext.Releases
+                        .Include(r => r.Publication)
+                        .AsAsyncEnumerable()
+                        .Where(r => r.PublicationId == release.PublicationId)
+                        .ToListAsync();
+
+                    var allLatestReleaseIds = allReleases
+                        .Where(r => r.Publication.IsLatestVersionOfRelease(r.Id))
+                        .Select(r => r.Id)
+                        .ToList();
+
+                    var allContributorReleaseRoles = await _contentDbContext.UserReleaseRoles
+                        .Include(usr => usr.User)
+                        .AsAsyncEnumerable()
+                        .Where(rr =>
+                            allLatestReleaseIds.Contains(rr.ReleaseId)
+                            && rr.Role == ReleaseRole.Contributor)
+                        .ToListAsync();
+
+                    var allPublicationContributors = allContributorReleaseRoles
```

Comment:
I don't think this is possible because I use `allContributorReleaseRoles` multiple times - if I skipped saving it as a variable, then I think I would need a db call for each user to get `var roleForThisRelease`.

I did wonder if we could add a ReleaseRoles field onto User.cs, so you could do
```
var allPublicationContributors = await _context.Users
    .Include(u => u.ReleaseRoles)
    ???
    .ToList();

....

var roleForThisRelease = user.ReleaseRoles.SingleOrDefault(releaseRole => releaseRole.ReleaseId == releaseId);
```

I've had a quick go at this, but I'm not sure it will actually save a DB call, so I'm going to drop this for now. I can return to it.

Reply:
I think there is something here, but it requires me to add `public List<UserReleaseRole> UserReleaseRoles { get; set; }` to `User.cs` and after I do that, half the unit tests fail. I don't want to spend the time investigating this right now, especially as the code below might not work and save a DB call. So I'm going to resolve this.

The code in case its useful in the future:
```
                    var allPublicationContributors = await _contentDbContext.Users
                        .Include(u => u.UserReleaseRoles)
                        .AsAsyncEnumerable()
                        .Where(user =>
                            user.UserReleaseRoles
                                .Select(urr => urr.Id)
                                .Intersect(allLatestReleaseIds).Any()
                            )
                        .ToListAsync();

                    return allPublicationContributors
                        .Select(user =>
                        {
                            var roleForThisRelease = user.UserReleaseRoles.SingleOrDefault(releaseRole =>
                                releaseRole.ReleaseId == releaseId);

                            return new ManageAccessPageContributorViewModel
                            {
                                UserId = user.Id,
                                UserFullName = user.DisplayName,
                                ReleaseId = releaseId,
                                // Some don't have a contributor release role for this release, as they are a
                                // contributor on a different release on the same publication
                                ReleaseRoleId = roleForThisRelease?.Id,
                            };
                        }).ToList();
```